src/ray/stats/metric.h

```diff
@@ -191,6 +201,152 @@ struct MetricPoint {
+                  const std::vector<opencensus::tags::TagKey> &tag_keys,
+                  const std::vector<double> &buckets) {
+  using I = StatsTypeMap<T>;
+  auto view_descriptor = opencensus::stats::ViewDescriptor()
+                             .set_name(name + I::val)
+                             .set_description(description)
+                             .set_measure(name)
+                             .set_aggregation(I::Aggregation(buckets));
+  details::RegisterAsView(view_descriptor, tag_keys);
+}
+
+template <typename T = void>
+void RegisterViewList(const std::string &name, const std::string &description,
+                      const std::vector<opencensus::tags::TagKey> &tag_keys,
+                      const std::vector<double> &buckets) {}
+
+template <StatsType T, StatsType... Ts>
+void RegisterViewList(const std::string &name, const std::string &description,
+                      const std::vector<opencensus::tags::TagKey> &tag_keys,
+                      const std::vector<double> &buckets) {
+  RegisterView<T>(name, description, tag_keys, buckets);
+  RegisterViewList<Ts...>(name, description, tag_keys, buckets);
+}
+
+inline std::vector<opencensus::tags::TagKey> convertTags(
+    const std::vector<std::string> &names) {
+  std::vector<opencensus::tags::TagKey> ret;
+  for (auto &n : names) {
+    ret.push_back(TagKeyType::Register(n));
+  }
+  return ret;
+}
+
+class Stats {
+  using Measure = opencensus::stats::Measure<double>;
+
+ public:
+  Stats(const std::string &measure, const std::string &description,
+        std::vector<std::string> tag_keys, std::vector<double> buckets,
+        std::function<void(const std::string &, const std::string,
+                           const std::vector<opencensus::tags::TagKey>,
+                           const std::vector<double> &buckets)>
+            register_func)
+      : tag_keys_(tag_keys) {
+    auto f = [register_func, measure, description, buckets, this]() {
+      measure_ = std::make_unique<Measure>(Measure::Register(measure, description, ""));
+      register_func(measure, description, convertTags(tag_keys_), buckets);
+    };
+
+    if (StatsConfig::instance().IsInitialized()) {
+      f();
+    } else {
+      StatsConfig::instance().AddInitializer(f);
+    }
+  }
+
+  void Record(double val) { Record(val, std::unordered_map<std::string, std::string>()); }
+
+  void Record(double val, std::string tag_val) {
+    RAY_CHECK(tag_keys_.size() == 1);
+    std::unordered_map<std::string, std::string> tags{{tag_keys_[0], std::move(tag_val)}};
+    Record(val, std::move(tags));
+  }
+
+  void Record(double val, std::unordered_map<std::string, std::string> tags) {
+    if (StatsConfig::instance().IsStatsDisabled() || !measure_) {
+      return;
+    }
+    TagsType combined_tags = StatsConfig::instance().GetGlobalTags();
+    // In case that tag containing non-printable chars
+    for (auto &t : tags) {
+      for (auto &c : t.second) {
```

Comment:
Can you comment what this block does?

Reply:
I think I commented at L318? Do you want me to add an extra message there?

Reply:
Yes! I think I didn't really understand the comment!